src/react/components/InterfaceEditor.tsx

```diff
@@ -0,0 +1,945 @@
+    if (parsedInterface.type === 'datastream' && parsedInterface.aggregation === 'object') {
+      setDatastreamOptions({
+        reliability: mapping?.reliability,
+        retention: mapping?.retention,
+        expiry: mapping?.expiry,
+        databaseRetentionPolicy: mapping?.databaseRetentionPolicy,
+        databaseRetentionTtl: mapping?.databaseRetentionTtl,
+        explicitTimestamp: mapping?.explicitTimestamp,
+      });
+    } else {
+      setDatastreamOptions({});
+    }
+  }, []);
+
+  useEffect(() => {
+    setInterfaceDraft((draft) => {
+      if (draft.aggregation !== 'object' || draft.mappings.length === 0) {
+        return draft;
+      }
+      const mappings = draft.mappings.map((mapping) => ({
+        ...mapping,
+        reliability: datastreamOptions.reliability,
+        retention: datastreamOptions.retention,
+        expiry: datastreamOptions.expiry,
+        databaseRetentionPolicy: datastreamOptions.databaseRetentionPolicy,
+        databaseRetentionTtl: datastreamOptions.databaseRetentionTtl,
+        explicitTimestamp: datastreamOptions.explicitTimestamp,
+      }));
+      return { ...draft, mappings };
+    });
+  }, [datastreamOptions]);
+
+  let interfaceValidationErrors: { [key: string]: string } = {};
+  try {
+    AstarteInterface.validation.validateSync(interfaceDraft, { abortEarly: false });
+  } catch (err) {
+    interfaceValidationErrors = _.mapValues(
+      _.keyBy(_.uniqBy(err.inner, 'path'), 'path'),
+      'message',
+    );
+  }
+  const isValidInterface = _.isEmpty(interfaceValidationErrors);
+
+  useEffect(() => {
+    const formattedInterfaceSource = formatJSON(AstarteInterface.toJSON(interfaceDraft));
+    setInterfaceSource((source) =>
+      formatJSONText(source) !== formattedInterfaceSource ? formattedInterfaceSource : source,
+    );
+    onChange(interfaceDraft, isValidInterface);
+  }, [interfaceDraft, isValidInterface]);
+
+  const interfaceValidationWarnings = computeInterfaceWarnings(interfaceDraft);
+
+  let isValidInterfaceSource = true;
+  let interfaceSourceError = '';
+  if (!checkValidJSONText(interfaceSource)) {
+    isValidInterfaceSource = false;
+    interfaceSourceError = 'Invalid JSON';
+  } else {
+    try {
+      parseAstarteInterfaceJSON(JSON.parse(interfaceSource));
+    } catch (err) {
+      isValidInterfaceSource = false;
+      interfaceSourceError = err.message;
+    }
+  }
+
+  const mappingToEdit = _.nth(interfaceDraft.mappings, mappingToEditIndex);
+
+  const showInterfaceExpiry =
+    datastreamOptions.retention === 'volatile' || datastreamOptions.retention === 'stored';
+
+  const showInterfaceDatabaseRetentionTtl = datastreamOptions.databaseRetentionPolicy === 'use_ttl';
+
+  return (
+    <Row>
+      <Col md={isSourceVisible ? 6 : 12}>
+        <Container fluid className="bg-white rounded p-3">
+          <Form>
+            <Form.Row className="mb-2">
+              <Col md={6}>
+                <Form.Group controlId="interfaceName">
```

Comment:
same as previous comment. This is a huge junk of jsx, it would be easier to read if you divide it in a few macro blocks

Reply:
I understand your point, but I have the same concern I wrote in the previous comment: I am not sure hiding details behind custom blocks would improve readability, it would probably increase the scrolling though